// @tsparticles/interaction-external-pop/umd/Popper.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@tsparticles/engine"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.Popper = void 0;
    const engine_1 = require("@tsparticles/engine");
    const popMode = "pop";
    class Popper extends engine_1.ExternalInteractorBase {
        constructor(container) {
            super(container);
            this.handleClickMode = (mode) => {
                const container = this.container;
                if (mode !== popMode) {
                    return;
                }
                const clickPos = container.interactivity.mouse.clickPosition;
                if (!clickPos) {
                    return;
                }
                const poppedParticles = container.particles.quadTree.queryCircle(clickPos, container.retina.pixelRatio);
                if (!poppedParticles.length) {
                    return;
                }
                for (const particle of poppedParticles) {
                    container.particles.remove(particle);
                }
            };
        }
        clear() {
        }
        init() {
        }
        interact() {
        }
        isEnabled() {
            return true;
        }
        reset() {
        }
    }
    exports.Popper = Popper;
});
