// @tsparticles/plugin-emitters-shape-canvas/umd/Options/Classes/TextFontOptions.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TextFontOptions = void 0;
    class TextFontOptions {
        constructor() {
            this.family = "Verdana";
            this.size = 32;
            this.style = "";
            this.variant = "";
            this.weight = "";
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.family !== undefined) {
                this.family = data.family;
            }
            if (data.size !== undefined) {
                this.size = data.size;
            }
            if (data.style !== undefined) {
                this.style = data.style;
            }
            if (data.variant !== undefined) {
                this.variant = data.variant;
            }
            if (data.weight !== undefined) {
                this.weight = data.weight;
            }
        }
    }
    exports.TextFontOptions = TextFontOptions;
});
