// @tsparticles/plugin-sounds/umd/Options/Classes/SoundsEvent.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@tsparticles/engine", "./SoundsAudio.js", "./SoundsMelody.js", "./SoundsNote.js"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.SoundsEvent = void 0;
    const engine_1 = require("@tsparticles/engine");
    const SoundsAudio_js_1 = require("./SoundsAudio.js");
    const SoundsMelody_js_1 = require("./SoundsMelody.js");
    const SoundsNote_js_1 = require("./SoundsNote.js");
    class SoundsEvent {
        constructor() {
            this.event = [];
            this.notes = [];
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.event !== undefined) {
                this.event = data.event;
            }
            if (data.audio !== undefined) {
                if ((0, engine_1.isArray)(data.audio)) {
                    this.audio = data.audio.map(s => {
                        const tmp = new SoundsAudio_js_1.SoundsAudio();
                        tmp.load(s);
                        return tmp;
                    });
                }
                else {
                    this.audio = new SoundsAudio_js_1.SoundsAudio();
                    this.audio.load(data.audio);
                }
            }
            if (data.notes !== undefined) {
                this.notes = data.notes.map(t => {
                    const tmp = new SoundsNote_js_1.SoundsNote();
                    tmp.load(t);
                    return tmp;
                });
            }
            if (data.melodies !== undefined) {
                this.melodies = data.melodies.map(t => {
                    const tmp = new SoundsMelody_js_1.SoundsMelody();
                    tmp.load(t);
                    return tmp;
                });
            }
            if (data.filter) {
                if ((0, engine_1.isString)(data.filter)) {
                    const filterFunc = window[data.filter];
                    if ((0, engine_1.isFunction)(filterFunc)) {
                        this.filter = filterFunc;
                    }
                }
                else {
                    this.filter = data.filter;
                }
            }
        }
    }
    exports.SoundsEvent = SoundsEvent;
});
