// @tsparticles/interaction-external-particle/umd/Options/Classes/InteractivityParticle.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@tsparticles/engine"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.InteractivityParticle = void 0;
    const engine_1 = require("@tsparticles/engine");
    class InteractivityParticle {
        constructor() {
            this.replaceCursor = false;
            this.pauseOnStop = false;
            this.stopDelay = 0;
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.options !== undefined) {
                this.options = (0, engine_1.deepExtend)({}, data.options);
            }
            if (data.replaceCursor !== undefined) {
                this.replaceCursor = data.replaceCursor;
            }
            if (data.pauseOnStop !== undefined) {
                this.pauseOnStop = data.pauseOnStop;
            }
            if (data.stopDelay !== undefined) {
                this.stopDelay = data.stopDelay;
            }
        }
    }
    exports.InteractivityParticle = InteractivityParticle;
});
