// @tsparticles/plugin-canvas-mask/umd/Options/Classes/TextMask.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./FontTextMask.js", "./TextMaskLine.js"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TextMask = void 0;
    const FontTextMask_js_1 = require("./FontTextMask.js");
    const TextMaskLine_js_1 = require("./TextMaskLine.js");
    class TextMask {
        constructor() {
            this.color = "#000000";
            this.font = new FontTextMask_js_1.FontTextMask();
            this.lines = new TextMaskLine_js_1.TextMaskLine();
            this.text = "";
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.color !== undefined) {
                this.color = data.color;
            }
            this.font.load(data.font);
            this.lines.load(data.lines);
            if (data.text !== undefined) {
                this.text = data.text;
            }
        }
    }
    exports.TextMask = TextMask;
});
