// @tsparticles/plugin-sounds/cjs/Options/Classes/Sounds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sounds = void 0;
const SoundsEvent_js_1 = require("./SoundsEvent.js");
const SoundsIcons_js_1 = require("./SoundsIcons.js");
const SoundsVolume_js_1 = require("./SoundsVolume.js");
class Sounds {
    constructor() {
        this.autoPlay = true;
        this.enable = false;
        this.events = [];
        this.icons = new SoundsIcons_js_1.SoundsIcons();
        this.volume = new SoundsVolume_js_1.SoundsVolume();
    }
    load(data) {
        if (!data) {
            return;
        }
        if (data.autoPlay !== undefined) {
            this.autoPlay = data.autoPlay;
        }
        if (data.enable !== undefined) {
            this.enable = data.enable;
        }
        if (data.events !== undefined) {
            this.events = data.events.map(t => {
                const event = new SoundsEvent_js_1.SoundsEvent();
                event.load(t);
                return event;
            });
        }
        this.icons.load(data.icons);
        if (data.volume !== undefined) {
            this.volume.load(data.volume);
        }
    }
}
exports.Sounds = Sounds;
