// @tsparticles/plugin-emitters-shape-polygon/umd/EmittersPolygonShape.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@tsparticles/plugin-emitters", "@tsparticles/engine", "./utils.js"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.EmittersPolygonShape = void 0;
    const plugin_emitters_1 = require("@tsparticles/plugin-emitters");
    const engine_1 = require("@tsparticles/engine");
    const utils_js_1 = require("./utils.js");
    const half = 0.5;
    class EmittersPolygonShape extends plugin_emitters_1.EmitterShapeBase {
        constructor(position, size, fill, options) {
            super(position, size, fill, options);
            this.sides = options.sides;
            this.angle = (0, engine_1.degToRad)(options.angle);
            this.polygon = (0, utils_js_1.generateRandomPolygon)(position, this.sides, size.width * half, this.angle);
        }
        async init() {
        }
        randomPosition() {
            const fill = this.fill, polygon = this.polygon, res = fill ? (0, utils_js_1.generateRandomPointWithinPolygon)(polygon) : (0, utils_js_1.generateRandomPointOnPolygonPerimeter)(polygon);
            return res ? { position: res } : null;
        }
        resize(position, size) {
            super.resize(position, size);
            this.polygon = (0, utils_js_1.generateRandomPolygon)(position, this.sides, size.width * half, this.angle);
        }
    }
    exports.EmittersPolygonShape = EmittersPolygonShape;
});
