// @tsparticles/updater-gradient/umd/Options/Classes/GradientColorOpacityAnimation.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@tsparticles/engine"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.GradientColorOpacityAnimation = void 0;
    const engine_1 = require("@tsparticles/engine");
    class GradientColorOpacityAnimation {
        constructor() {
            this.count = 0;
            this.enable = false;
            this.speed = 0;
            this.decay = 0;
            this.delay = 0;
            this.sync = false;
            this.startValue = engine_1.StartValueType.random;
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.count !== undefined) {
                this.count = (0, engine_1.setRangeValue)(data.count);
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.speed !== undefined) {
                this.speed = (0, engine_1.setRangeValue)(data.speed);
            }
            if (data.sync !== undefined) {
                this.sync = data.sync;
            }
            if (data.startValue !== undefined) {
                this.startValue = data.startValue;
            }
            if (data.decay !== undefined) {
                this.decay = (0, engine_1.setRangeValue)(data.decay);
            }
            if (data.delay !== undefined) {
                this.delay = (0, engine_1.setRangeValue)(data.delay);
            }
        }
    }
    exports.GradientColorOpacityAnimation = GradientColorOpacityAnimation;
});
