// @tsparticles/plugin-poisson-disc/umd/Options/Classes/Poisson.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.Poisson = void 0;
    class Poisson {
        constructor() {
            this.enable = false;
            this.dimensions = 2;
            this.radius = 0;
            this.retries = 30;
            this.steps = 0;
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.dimensions !== undefined) {
                this.dimensions = data.dimensions;
            }
            if (data.radius !== undefined) {
                this.radius = data.radius;
            }
            if (data.retries !== undefined) {
                this.retries = data.retries;
            }
        }
    }
    exports.Poisson = Poisson;
});
